sdk/servicebus/service-bus/review/service-bus.api.md

```diff
@@ -260,19 +260,43 @@ export class ServiceBusClient {
     constructor(connectionString: string, options?: ServiceBusClientOptions);
     constructor(fullyQualifiedNamespace: string, credential: TokenCredential, options?: ServiceBusClientOptions);
     close(): Promise<void>;
-    createDeadLetterReceiver(queueName: string, receiveMode: "peekLock"): Receiver<ReceivedMessageWithLock>;
-    createDeadLetterReceiver(queueName: string, receiveMode: "receiveAndDelete"): Receiver<ReceivedMessage>;
-    createDeadLetterReceiver(topicName: string, subscriptionName: string, receiveMode: "peekLock"): Receiver<ReceivedMessageWithLock>;
-    createDeadLetterReceiver(topicName: string, subscriptionName: string, receiveMode: "receiveAndDelete"): Receiver<ReceivedMessage>;
-    createReceiver(queueName: string, receiveMode: "peekLock"): Receiver<ReceivedMessageWithLock>;
-    createReceiver(queueName: string, receiveMode: "receiveAndDelete"): Receiver<ReceivedMessage>;
-    createReceiver(topicName: string, subscriptionName: string, receiveMode: "peekLock"): Receiver<ReceivedMessageWithLock>;
-    createReceiver(topicName: string, subscriptionName: string, receiveMode: "receiveAndDelete"): Receiver<ReceivedMessage>;
+    createDeadLetterReceiver(queueName: string, options?: {
+        receiveMode?: "peekLock";
+    }): Receiver<ReceivedMessageWithLock>;
+    createDeadLetterReceiver(queueName: string, options?: {
+        receiveMode?: "receiveAndDelete";
+    }): Receiver<ReceivedMessage>;
```

Comment:
How does this show up in the editor when one types `createDeadLe`?
Do they get 4 suggestions? 2 for queue, 2 for topic/subscription?
Would the 2 for queue show up any different? Or both will have the same input signature with different outputs?

Reply:
When you type `createDeadLe`, you'll only see one suggestion since these are overloads.
When you type `createDeadLetterReceiver`, you'll be seeing 4 overloads.
![image](https://user-images.githubusercontent.com/10452642/88095673-57d7cb80-cb4a-11ea-9352-076eeddfb4d3.png)
![image](https://user-images.githubusercontent.com/10452642/88095616-4262a180-cb4a-11ea-9370-bb8613584803.png)

Reply:
And we got lucky that the peeklock one shows up first because p comes before r

Reply:
No, it's the order of the overloads.
We need to put the "peekLock" overloads before the "receiveAndDelete" ones.

Reply:
This is the first time we will be having overloads for different values of a property in the options bag.

One of the reasons behind making `receiveMode` optional was to avoid the situation where the user has to know the difference between the two modes and make a decision. This applies to everything that we consider as "optional". By making overloads for optional properties, we bring that decision front and center to the user again.

cc @richardpark-msft, @chradek

Reply:
I see the point. Unfortunately, I don't see a way other than this if we want to have different return types for the receiver in different modes. 

With this, the samples/readme snippets would not have the receiveMode param passed to the createReceiver(). For such getting started scenarios, users won't have to know the receiveMode and can get away with the defaults.

----------------
The other way(to avoid the overloads) was to use the conditional types(e663dfc) which would have only 2 overloads, one for queues, the other for subscriptions.
![image](https://user-images.githubusercontent.com/10452642/88640229-8be34d00-d072-11ea-859f-50a5abc6a221.png)
![image](https://user-images.githubusercontent.com/10452642/88640370-ba612800-d072-11ea-84f8-eed95cfd6593.png)

Based on the current input, the current/first signature gets updated(dynamic). The only concern is that the _other_ signatures would be unreadable.

Reply:
Ramya and I like the second option if we can make the docs look good (e.g. we shouldn't see a conditional type in the return signature if at all possible).

Reply:
I do love the second option, I don't think I have the control over the signature rendered in VSCode.
Also, more info... when I said the signature gets updated dynamically, this is what I meant.

Current:
![image](https://user-images.githubusercontent.com/10452642/89003375-72314800-d2b4-11ea-8331-9d6f10de5154.png)
Other:
![image](https://user-images.githubusercontent.com/10452642/89003424-92f99d80-d2b4-11ea-8086-16076a4afac9.png)

Reply:
TYPEDOC for the conditional types.
![image](https://user-images.githubusercontent.com/10452642/89250934-b5e4c400-d5ca-11ea-8b05-208682dcc095.png)

Reply:
More info/discussions - #10034